service/lib/agama/software/callbacks/signature.rb

```diff
@@ -109,17 +99,88 @@ def import_gpg_key(key, _repo_id)
           )
 
           questions_client.ask(question) do |question_client|
-            question_client.answer == :Trust
+            question_client.answer == trust_label.to_sym
+          end
+        end
+
+        # Callback to handle unknown GPG keys
+        #
+        # @param filename [String] Name of the file.
+        # @param key_id [String] Key ID.
+        # @param repo_id [String] Repository ID.
+        def accept_unknown_gpg_key(filename, key_id, repo_id)
+          source = build_source(filename, repo_id)
+          message = format(
+            _("%{source} is digitally signed with the following unknown GnuPG key: " \
+              "%{key_id}. Use it anyway?"),
+            source: source, key_id: key_id
+          )
+
+          question = Agama::Question.new(
+            qclass:         "software.unknown_gpg",
+            text:           message,
+            options:        [yes_label.to_sym, no_label.to_sym],
+            default_option: no_label.to_sym,
+            data:           {
+              "id"       => key_id,
+              "filename" => filename
+            }
+          )
+
+          questions_client.ask(question) do |question_client|
+            question_client.answer == yes_label.to_sym
+          end
+        end
+
+        # Callback to handle file verification failures
+        #
+        # @param filename [String] File name
+        # @param key [Hash] GPG key data (id, name, fingerprint, etc.)
+        # @param repo_id [Integer] Repository ID
+        def accept_verification_failed(filename, key, repo_id)
+          source = build_source(filename, repo_id)
+          message = format(
+            _("%{source} is signed with the following GnuPG key, but the integrity check failed: " \
+              "%{key_id} (%{key_name}). Use it anyway?"),
+            source: source, key_id: key["id"], key_name: key["name"]
+          )
+
+          question = Agama::Question.new(
+            qclass:         "software.unsigned_file",
+            text:           message,
+            options:        [yes_label.to_sym, no_label.to_sym],
+            default_option: no_label.to_sym,
+            data:           { "filename" => filename }
+          )
+          questions_client.ask(question) do |question_client|
+            question_client.answer == yes_label.to_sym
           end
         end
 
       private
 
-        # @return [Agama::DBus::Clients::Questions]
-        attr_reader :questions_client
+        # Builds the file source description
+        #
+        # @param filename [String] File name
+        # @param repo_id [Integer] Repository ID
+        # @return [String]
+        def build_source(filename, repo_id)
+          repo = Yast::Pkg.SourceGeneralData(repo_id)
+          if repo
+            format(
+              _("The file %{filename} from repository %{repo_name} (%{repo_url})"),
```

Comment:
As we do not give the repositories any meaningful names and just use the URL as fallback I'd omit the repository name in the message.